### ISSUE TITLE
Controller getting re-instantiated on render

### ISSUE THREAD

**Issue by a-s-o** [lhorie/mithril.js#803]
Hello,

I am stuck on something simple; started a discussion in Gitter but did not receive an answer.

[The docs say](http://lhorie.github.io/mithril/mithril.component.html#rendering-components):

> There are three ways to render a component:
> 
> * m.route (if you are building a single-page application that has multiple pages)
> * m.mount (if your app only has one page)
> * m.render (if you are integrating Mithril's rendering engine into a larger framework and wish to manage redrawing yourself).
> 
> The controller function is called once when the component is rendered. Subsequently, the view function is called and will be called again anytime a redraw is required. The return value of the controller function is passed to the view as its first argument.

Having said that if I do the following, I get multiple controller instantiations. What am I missing?

```javascript
let count = 0;
const component = {
  controller () {
    count++;
  },
  view () {
    return m('h3', `Controller was re-instantiated ${count} times`);
  }
};

setInterval(function () {
  m.render(document.body, component);
}, 1000);
```
[Codepen](http://codepen.io/anon/pen/epBGYy?editors=101)

**Comment by Zhouzi** [comment]
> The controller function is called once when the component is rendered

`m.render` calls the controller function and it's wrapped within a `setInterval` so it's basically re-instantiating the controller every `1000` ms or so.

**Comment by a-s-o** [comment]
Hey thanks for taking a look.

I understand that the render is being invoked over and over in the interval, but `m.render` should be creating a controller once on the first render, then on all future renders based on my understanding, the element should get `diff`ed and only the component view should be called. Do I have that wrong?

Except of course on route changes, or `m.redraw.strategy('all')` or after a forced render using the third argument to `m.render`; all of which I am not doing in the above example.

**Comment by codeclown** [comment]
I might be missing the point here, but why aren't you using `m.mount`?

```js
let count = 0;

const component = {
  controller () {
    setInterval(function () {
      m.startComputation();
      count++;
      m.endComputation();
    }, 1000);
  },
  view () {
    return m('h3', `Controller was re-instantiated ${count} times`);
  }
};

m.mount(document.body, component);
```

**Comment by barneycarroll** [comment]
The methods described in that passage are ways of instantiating a component. Each time you call them they will (re-)initialise the components passed in (actually, that won't happen with `m.route` – this will just change the mappings).

When it says:

> …the view function is called and will be called again anytime a redraw is required

…this is not referring to repeat component invocations: it is referring to explicit or implicit calls to `m.redraw`. Furthermore, `m.render` is not intended to be used for redrawing components – the documentation is arguably misleading on this point: if you intend for the passed in component to be redrawn, you should use `m.mount` as @codeclown says; `m.render` is designed for situations where you want to explicitly control instantiation and redraw. However his modified code is confusing, because `count` is incremented every second regardless of instantiations.

This might make things clearer:

```javascript
let redrawCount = 0;
let instanceCount = 0;

const component = {
  controller () {
      instanceCount++;
  },
  view () {
    return m('h3', `Controller was re-instantiated ${instanceCount} times, redrawn ${redrawCount++} times`);
  }
};

m.mount(document.body, component);

setInterval(m.redraw, 1000);
```

[Demo here](http://jsbin.com/kizuso/edit?html,js,output).

**Comment by a-s-o** [comment]
Hi, thanks @codeclown  and @barneycarroll 

You are right about using `m.mount`; actually to explain my use case, I ran into this issue when trying to write some integration tests. I have a series of nested components and I was writing tests to check whether the `config` function for a particular element/component are called in a predictable manner. I.e. call with `isInitialized = false` on first render and then `isInitialized = true` on subsequent redraws. 

I was using `m.render` to do this and what ended up happening was I saw that `isInitialized` was always `false` which let to me noticing that the controllers were being re-instantiated in the tests as well. Finally, as Barney noted, I turned to the docs and was confused. Maybe the excerpt included above needs to be clarified.

I will give it another shot with `m.mount`; curiously using the same mock `window` used in mithril tests leads to the same even with `m.mount` but that may be a mock issue. I will report back shortly. 

Barney, thanks for the demo
